src/nbl/video/utilities/CScanner.cpp

```diff
@@ -1,52 +1,52 @@
 #include "nbl/video/utilities/CScanner.h"
 
-using namespace nbl;
-using namespace video;
+namespace nbl::video
+{
+    
+asset::ICPUShader* CScanner::getDefaultShader(const E_SCAN_TYPE scanType, const E_DATA_TYPE dataType, const E_OPERATOR op, const uint32_t scratchSz)
+{
+    if (!m_shaders[scanType][dataType][op])
+        m_shaders[scanType][dataType][op] = createShader(false,scanType,dataType,op,scratchSz);
+    return m_shaders[scanType][dataType][op].get();
+}
 
-#if 0 // TODO: port
-core::smart_refctd_ptr<asset::ICPUShader> CScanner::createShader(const bool indirect, const E_SCAN_TYPE scanType, const E_DATA_TYPE dataType, const E_OPERATOR op) const
+IGPUShader* CScanner::getDefaultSpecializedShader(const E_SCAN_TYPE scanType, const E_DATA_TYPE dataType, const E_OPERATOR op, const uint32_t scratchSz)
 {
-	auto system = m_device->getPhysicalDevice()->getSystem();
-	core::smart_refctd_ptr<const system::IFile> glsl;
-	{
-		auto loadBuiltinData = [&](const std::string _path) -> core::smart_refctd_ptr<const nbl::system::IFile>
-		{
-			nbl::system::ISystem::future_t<core::smart_refctd_ptr<nbl::system::IFile>> future;
-			system->createFile(future, system::path(_path), core::bitflag(nbl::system::IFileBase::ECF_READ) | nbl::system::IFileBase::ECF_MAPPABLE);
-			if (future.wait())
-				return future.copy();
-			return nullptr;
-		};
+    if (!m_specialized_shaders[scanType][dataType][op])
+    {
+        auto cpuShader = core::smart_refctd_ptr<asset::ICPUShader>(getDefaultShader(scanType,dataType,op,scratchSz));
+        cpuShader->setFilePathHint("nbl/builtin/hlsl/scan/direct.hlsl");
+        cpuShader->setShaderStage(asset::IShader::ESS_COMPUTE);
+
+        auto gpushader = m_device->createShader(cpuShader.get());
+
+        m_specialized_shaders[scanType][dataType][op] = gpushader;
+    }
+    return m_specialized_shaders[scanType][dataType][op].get();
+}
 
-		if(indirect)
```

Comment:
these 3 utilities are practically identical for CScan and CReduce, why not hoist them to CArithmeticOps?